File: fastpay-assistant-ml/fastpay_assistant/ml/action_ranker.py
from __future__ import annotations

from dataclasses import dataclass

from fastpay_assistant.ml.config import (
    INTENT_MATCH_BOOST,
    RISK_FLAG_BOOST,
    ROUTE_VISIT_PENALTY,
    ROUTE_VISIT_PENALTY_AFTER,
)
from fastpay_assistant.types import AssistantIntent, ChatAction, UserProfile


@dataclass
class RankedAction:
    action: ChatAction
    score: float


# intent → preferred routes
INTENT_ROUTES: dict[AssistantIntent, list[tuple[str, str]]] = {
    AssistantIntent.BALANCE: [("Open Wallet", "/wallet")],
    AssistantIntent.CASH_FLOW: [("Review cash flow", "/analytics")],
    AssistantIntent.PLANNING: [
        ("Open Analytics", "/analytics"),
        ("Add funds", "/buy"),
    ],
    AssistantIntent.BUDGET: [("View goals", "/analytics?mode=goals")],
    AssistantIntent.KYC: [("Complete KYC", "/(auth)/kyc")],
    AssistantIntent.PASSCODE: [("Reset passcode", "/forgot-passcode")],
    AssistantIntent.NAVIGATE: [],
    AssistantIntent.PRODUCT_HELP: [
        ("Open Bills", "/bills"),
        ("Transfer", "/wallet/transfer"),
    ],
    AssistantIntent.GENERAL: [("Open Support", "/support")],
    AssistantIntent.EXTERNAL_INFO: [("Open Settings", "/settings")],
}

RISK_ACTIONS: dict[str, list[tuple[str, str]]] = {
    "high_spend": [("Review Analytics", "/analytics")],
    "low_savings": [("Set savings goal", "/analytics?mode=goals"), ("Add funds", "/buy")],
    "loan_interest": [("Loan info", "/loan/apply")],
}


def _routes_match(a: str, b: str) -> bool:
    return a == b or a.split("?")[0] == b.split("?")[0]


def dedupe_actions(actions: list[ChatAction]) -> list[ChatAction]:
    seen: set[str] = set()
    out: list[ChatAction] = []
    for action in actions:
        if action.href in seen:
            continue
        seen.add(action.href)
        out.append(action)
    return out


def rank_actions(
    intent: AssistantIntent,
    profile: UserProfile | None,
    candidates: list[ChatAction] | None = None,
    *,
    top_n: int = 2,
) -> list[ChatAction]:
    scored: list[RankedAction] = []
    visits = profile.route_visit_counts if profile else {}

    pool: list[ChatAction] = list(candidates or [])
    for label, href in INTENT_ROUTES.get(intent, []):
        pool.append(ChatAction(label=label, href=href))

    if profile:
        for flag in profile.risk_flags:
            # never push loans when low_savings unless user intent is loan-related
            if flag == "loan_interest" and "low_savings" in profile.risk_flags:
                if intent != AssistantIntent.PRODUCT_HELP:
                    continue
            for label, href in RISK_ACTIONS.get(flag, []):
                if href == "/loan/apply" and "low_savings" in profile.risk_flags:
                    continue
                pool.append(ChatAction(label=label, href=href))

    for action in dedupe_actions(pool):
        score = 0.0
        preferred = {h for _, h in INTENT_ROUTES.get(intent, [])}
        if any(_routes_match(action.href, p) for p in preferred):
            score += INTENT_MATCH_BOOST

        if profile:
            matched_flags: set[str] = set()
            for flag in profile.risk_flags:
                flag_hrefs = {h for _, h in RISK_ACTIONS.get(flag, [])}
                if action.href in flag_hrefs:
                    matched_flags.add(flag)
            score += RISK_FLAG_BOOST * len(matched_flags)

            count = visits.get(action.href, visits.get(action.href.split("?")[0], 0))
            if count >= ROUTE_VISIT_PENALTY_AFTER:
                score -= ROUTE_VISIT_PENALTY

        scored.append(RankedAction(action=action, score=score))

    scored.sort(key=lambda r: r.score, reverse=True)
    return [r.action for r in scored[:top_n]]

File: FastPay/lib/assistant/ml/action-ranker.ts
import type { ChatAction } from "@/lib/api/chat";

import type { AssistantIntent, UserProfile } from "../types";

import {
  INTENT_MATCH_BOOST,
  RISK_FLAG_BOOST,
  ROUTE_VISIT_PENALTY,
  ROUTE_VISIT_PENALTY_AFTER,
} from "./config";

const INTENT_ROUTES: Partial<
  Record<AssistantIntent, { label: string; href: string }[]>
> = {
  balance: [{ label: "Open Wallet", href: "/wallet" }],
  cash_flow: [{ label: "Review cash flow", href: "/analytics" }],
  planning: [
    { label: "Open Analytics", href: "/analytics" },
    { label: "Add funds", href: "/buy" },
  ],
  budget: [{ label: "View goals", href: "/analytics?mode=goals" }],
  kyc: [{ label: "Complete KYC", href: "/(auth)/kyc" }],
  passcode: [{ label: "Reset passcode", href: "/forgot-passcode" }],
  product_help: [
    { label: "Open Bills", href: "/bills" },
    { label: "Transfer", href: "/wallet/transfer" },
  ],
  general: [{ label: "Open Support", href: "/support" }],
  external_info: [{ label: "Open Settings", href: "/settings" }],
};

const RISK_ACTIONS: Record<string, { label: string; href: string }[]> = {
  high_spend: [{ label: "Review Analytics", href: "/analytics" }],
  low_savings: [
    { label: "Set savings goal", href: "/analytics?mode=goals" },
    { label: "Add funds", href: "/buy" },
  ],
  loan_interest: [{ label: "Loan info", href: "/loan/apply" }],
};

function routesMatch(a: string, b: string): boolean {
  return a === b || a.split("?")[0] === b.split("?")[0];
}

export function dedupeActions(actions: ChatAction[]): ChatAction[] {
  const seen = new Set<string>();
  return actions.filter((action) => {
    if (seen.has(action.href)) {
      return false;
    }
    seen.add(action.href);
    return true;
  });
}

export function rankActions(
  intent: AssistantIntent,
  profile: UserProfile | undefined,
  candidates: ChatAction[] = [],
  topN = 2,
): ChatAction[] {
  const visits = profile?.routeVisitCounts ?? {};
  const pool: ChatAction[] = [...candidates];

  for (const route of INTENT_ROUTES[intent] ?? []) {
    pool.push(route);
  }

  if (profile) {
    for (const flag of profile.riskFlags) {
      if (flag === "loan_interest" && profile.riskFlags.includes("low_savings")) {
        if (intent !== "product_help") {
          continue;
        }
      }
      for (const route of RISK_ACTIONS[flag] ?? []) {
        if (
          route.href === "/loan/apply" &&
          profile.riskFlags.includes("low_savings")
        ) {
          continue;
        }
        pool.push(route);
      }
    }
  }

  const preferred = new Set(
    (INTENT_ROUTES[intent] ?? []).map((r) => r.href),
  );

  const scored = dedupeActions(pool).map((action) => {
    let score = 0;
    if ([...preferred].some((p) => routesMatch(action.href, p))) {
      score += INTENT_MATCH_BOOST;
    }

    if (profile) {
      const matchedFlags = new Set<string>();
      for (const flag of profile.riskFlags) {
        const flagHrefs = new Set(
          (RISK_ACTIONS[flag] ?? []).map((r) => r.href),
        );
        if (flagHrefs.has(action.href)) {
          matchedFlags.add(flag);
        }
      }
      score += RISK_FLAG_BOOST * matchedFlags.size;

      const count =
        visits[action.href] ?? visits[action.href.split("?")[0]] ?? 0;
      if (count >= ROUTE_VISIT_PENALTY_AFTER) {
        score -= ROUTE_VISIT_PENALTY;
      }
    }

    return { action, score };
  });

  scored.sort((a, b) => b.score - a.score);
  return scored.slice(0, topN).map((s) => s.action);
}
